### PR TITLE
Introduce more checks for Declarative DSL: reassignment and block order

### PR DESCRIPTION
**Pull request by h0tk3y**
Fixes https://github.com/gradle/gradle-private/issues/4166

### CODE REVIEW

**Comment by h0tk3y** [comment]
@bot-gradle test this

**Comment by h0tk3y** [comment]
@bot-gradle test this

File: platforms/core-configuration/declarative-dsl-provider/src/integTest/groovy/org/gradle/internal/declarativedsl/settings/DeclarativeDslProjectSettingsIntegrationSpec.groovy [hunk 1]
```diff
@@ -86,6 +87,54 @@ class DeclarativeDslProjectSettingsIntegrationSpec extends AbstractIntegrationSp
...
+        when:
+        def failure = fails(":projects")
+
+        then:
+        failure.assertHasErrorOutput(errorMessage)
+
+        where:
+        kind               | content                                                                             | errorMessage
+        'plugins'          | 'pluginManagement { }\nplugins { }\nrootProject.name = "foo"\nplugins { }'          | ""
+        'pluginManagement' | 'pluginManagement { }\nplugins { }\nrootProject.name = "foo"\npluginManagement { }' | ""
```
Comment (eskatos) [inline platforms/core-configuration/declarative-dsl-provider/src/integTest/groovy/org/gradle/internal/declarativedsl/settings/DeclarativeDslProjectSettingsIntegrationSpec.groovy:120]: ❓ Is the empty error message intended?

Comment (h0tk3y) [reply platforms/core-configuration/declarative-dsl-provider/src/integTest/groovy/org/gradle/internal/declarativedsl/settings/DeclarativeDslProjectSettingsIntegrationSpec.groovy]: Nope, it was an oversight. Fixed in [`73940b1` (#29028)](https://github.com/gradle/gradle/pull/29028/commits/73940b15ca7b4b79ce525af2113726591d61e785).

**Review by eskatos — APPROVED** [review]

**Pull request merged**